src/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs

```diff
@@ -75,6 +77,40 @@ public class SslStreamNetworkStreamTest
             listener.Stop();
         }
 
+        [Fact]
+        [OuterLoop] // Test hits external azure server.
+        public async Task SslStream_NetworkStream_Renegotiation_Succeeds()
+        {
+            Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            await s.ConnectAsync(Configuration.Security.TlsRenegotiationServer, 443);
+            using (NetworkStream ns = new NetworkStream(s))
+            using (SslStream ssl = new SslStream(ns, true))
+            {
+                X509CertificateCollection certBundle = new X509CertificateCollection();
+                certBundle.Add(Configuration.Certificates.GetClientCertificate());
+
+                // Perform handshake to establish secure connection.
+                await ssl.AuthenticateAsClientAsync(Configuration.Security.TlsRenegotiationServer, certBundle, false);
+                Assert.True(ssl.IsAuthenticated);
+                Assert.True(ssl.IsEncrypted);
+
+                // Issue request that triggers regotiation from server.
+                byte[] message = Encoding.UTF8.GetBytes("GET /EchoClientCertificate.ashx HTTP/1.1\r\nHost: corefx-net-tls.azurewebsites.net\r\n\r\n");
+                await ssl.WriteAsync(message, 0, message.Length);
+
+                // Initiate Read operation, that results in starting renegotiation as per server response to the above request.
+                int bytesRead = await ssl.ReadAsync(message, 0, message.Length);
+
+                // SslStream.ReadAsync should return -1 bytes, indicating renegotiation operation is in progress.
+                Assert.Equal(-1, bytesRead);
```

Comment:
Are we sure this is always true?  What happens if the renegotiation is very fast?  This seems like a possible place where the test could break due to race/timing issues with the remote end point and the intervening network.  So, we should keep an eye on this for test failures over time.

Reply:
This is always true as long as the renegotiation codepath is hit, if the server changes it's behavior then this might break.. Even if it is pretty fast, we always return -1 for read operation if renegotiation status is received from the server. And only with subsequent read can the other data be consumed.